File: sdk-typescript/test/protect.test.ts
/**
 * Tests for AegisAgent TypeScript SDK — protect() (protect.ts)
 *
 * Uses node:test + node:http (no external deps) matching canon.test.ts style.
 * Mock gateway is a real loopback HTTP server so routing logic is exercised.
 *
 * Coverage:
 *   1. allow  → tool executes
 *   2. deny   → AegisDenied thrown, tool NOT run
 *   3. require_approval: pending→approved → poll, verify hash, consume, execute
 *   4. hash mismatch on authorize response → AegisHashMismatch, tool NOT run
 *   5. hash mismatch on poll status → AegisHashMismatch, tool NOT run
 *   6. consume fails 409 → error, tool NOT run
 *   7. gateway unreachable on mutating action → error, tool NOT run
 *   8. unknown decision → error (fail closed), tool NOT run
 *   9. approval rejected → error, tool NOT run
 *  10. missing approval info in require_approval → error (fail closed)
 */
import { test } from "node:test";
import assert from "node:assert/strict";
import http from "node:http";
import { AegisClient } from "../src/client.ts";
import { protect, AegisDenied, AegisHashMismatch } from "../src/protect.ts";
import type { ProtectOptions } from "../src/protect.ts";
import { canonicalHash } from "../src/canon.ts";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

interface ServerHandle {
  url: string;
  close: () => void;
}

function startServer(
  handler: (req: http.IncomingMessage, res: http.ServerResponse) => void
): Promise<ServerHandle> {
  return new Promise((resolve) => {
    const srv = http.createServer(handler);
    srv.listen(0, "127.0.0.1", () => {
      const addr = srv.address() as { port: number };
      resolve({ url: `http://127.0.0.1:${addr.port}`, close: () => srv.close() });
    });
  });
}

function serveJSON(res: http.ServerResponse, status: number, body: unknown): void {
  const json = JSON.stringify(body);
  res.writeHead(status, { "Content-Type": "application/json" });
  res.end(json);
}

function newClient(baseUrl: string): AegisClient {
  return new AegisClient({ baseUrl, agentToken: "tok", tenantId: "tid" });
}

/** Compute the action_hash the same way protect() does, for use in mock responses. */
function hashAction(
  tool: string,
  action: string,
  mutatesState: boolean,
  parameters: Record<string, unknown>,
  resource?: string
): string {
  return canonicalHash({
    tool,
    action,
    resource: resource ?? null,
    mutates_state: mutatesState,
    parameters,
  });
}

// Polling options with zero delay so tests run fast.
const fastOpts: ProtectOptions = { pollIntervalMs: 0, maxPolls: 5 };

// ---------------------------------------------------------------------------
// 1. allow → tool executes
// ---------------------------------------------------------------------------

test("protect — allow: tool is executed and return value forwarded", async () => {
  const { url, close } = await startServer((_req, res) => {
    serveJSON(res, 200, { decision: "allow", action_hash: "any" });
  });
  try {
    const client = newClient(url);
    let called = false;
    const result = await protect(
      client,
      { tool: "github", action: "list_prs", mutatesState: false, parameters: { repo: "x" } },
      async () => {
        called = true;
        return "ok";
      },
      fastOpts
    );
    assert.ok(called, "tool must be called on allow");
    assert.equal(result, "ok");
  } finally {
    close();
  }
});

// ---------------------------------------------------------------------------
// 2. deny → AegisDenied thrown, tool NOT run
// ---------------------------------------------------------------------------

test("protect — deny: throws AegisDenied, tool NOT executed", async () => {
  const { url, close } = await startServer((_req, res) => {
    serveJSON(res, 200, { decision: "deny", reason: "forbidden by policy" });
  });
  try {
    const client = newClient(url);
    let called = false;
    await assert.rejects(
      () =>
        protect(
          client,
          { tool: "github", action: "force_push", mutatesState: true, parameters: { branch: "main" } },
          async () => {
            called = true;
            return "should not run";
          },
          fastOpts
        ),
      (err: unknown) => {
        assert.ok(err instanceof AegisDenied, `expected AegisDenied, got ${err}`);
        return true;
      }
    );
    assert.ok(!called, "tool must NOT be called on deny");
  } finally {
    close();
  }
});

// ---------------------------------------------------------------------------
// 3. require_approval: pending→approved → executes after consume
// ---------------------------------------------------------------------------

test("protect — approval: pending→approved polls, verifies hash, consumes, executes", async () => {
  const tool = "github";
  const action = "merge_pr";
  const params = { pr: 42 };
  const mutates = true;
  const approvalId = "appr-poll-001";
  const actionHash = hashAction(tool, action, mutates, params);

  let pollCount = 0;

  const { url, close } = await startServer((req, res) => {
    if (req.method === "POST" && req.url === "/v1/authorize") {
      serveJSON(res, 200, {
        decision: "require_approval",
        action_hash: actionHash,
        approval: { approval_id: approvalId, action_hash: actionHash },
      });
    } else if (req.method === "GET" && req.url === `/v1/approvals/${approvalId}`) {
      pollCount++;
      if (pollCount < 2) {
        serveJSON(res, 200, { status: "PENDING", action_hash: actionHash });
      } else {
        serveJSON(res, 200, { status: "APPROVED", action_hash: actionHash });
      }
    } else if (req.method === "POST" && req.url === `/v1/approvals/${approvalId}/consume`) {
      serveJSON(res, 200, { action_hash: actionHash });
    } else {
      res.writeHead(404);
      res.end("unexpected");
    }
  });

  try {
    const client = newClient(url);
    let called = false;
    await protect(
      client,
      { tool, action, mutatesState: mutates, parameters: params },
      async () => {
        called = true;
        return "done";
      },
      fastOpts
    );
    assert.ok(called, "tool must be called after approval");
    assert.ok(pollCount >= 2, "must have polled at least twice (pending then approved)");
  } finally {
    close();
  }
});

// ---------------------------------------------------------------------------
// 4. hash mismatch on authorize approval info → AegisHashMismatch, tool NOT run
// ---------------------------------------------------------------------------

test("protect — hash mismatch on authorize: throws AegisHashMismatch, tool NOT run", async () => {
  const badHash = "0".repeat(64);
  const approvalId = "appr-mismatch-001";

  const { url, close } = await startServer((_req, res) => {
    serveJSON(res, 200, {
      decision: "require_approval",
      action_hash: badHash,
      approval: { approval_id: approvalId, action_hash: badHash }, // wrong hash
    });
  });

  try {
    const client = newClient(url);
    let called = false;
    await assert.rejects(
      () =>
        protect(
          client,
          { tool: "github", action: "delete_branch", mutatesState: true, parameters: { branch: "x" } },
          async () => {
            called = true;
            return "";
          },
          fastOpts
        ),
      (err: unknown) => {
        assert.ok(err instanceof AegisHashMismatch, `expected AegisHashMismatch, got ${err}`);
        return true;
      }
    );
    assert.ok(!called, "tool must NOT be called on hash mismatch");
  } finally {
    close();
  }
});

// ---------------------------------------------------------------------------
// 5. hash mismatch on poll status (approve-then-swap attack) → refuse
// ---------------------------------------------------------------------------

test("protect — hash mismatch on poll APPROVED status: throws AegisHashMismatch, tool NOT run", async () => {
  const tool = "github";
  const action = "deploy";
  const params = { env: "prod" };
  const mutates = true;
  const approvalId = "appr-pollmismatch";
  const correctHash = hashAction(tool, action, mutates, params);
  const tamperedHash = "a".repeat(64);

  const { url, close } = await startServer((req, res) => {
    if (req.method === "POST" && req.url === "/v1/authorize") {
      serveJSON(res, 200, {
        decision: "require_approval",
        action_hash: correctHash,
        approval: { approval_id: approvalId, action_hash: correctHash },
      });
    } else if (req.method === "GET" && req.url === `/v1/approvals/${approvalId}`) {
      // Simulate approve-then-swap: gateway returns APPROVED with tampered hash
      serveJSON(res, 200, { status: "APPROVED", action_hash: tamperedHash });
    } else {
      res.writeHead(404);
      res.end("unexpected");
    }
  });

  try {
    const client = newClient(url);
    let called = false;
    await assert.rejects(
      () =>
        protect(
          client,
          { tool, action, mutatesState: mutates, parameters: params },
          async () => {
            called = true;
            return "";
          },
          fastOpts
        ),
      (err: unknown) => {
        assert.ok(err instanceof AegisHashMismatch, `expected AegisHashMismatch, got ${err}`);
        return true;
      }
    );
    assert.ok(!called, "tool must NOT be called when poll action_hash mismatches");
  } finally {
    close();
  }
});

// ---------------------------------------------------------------------------
// 6. consume fails 409 → error, tool NOT run (replay defence)
// ---------------------------------------------------------------------------

test("protect — consume 409: refuses and does NOT execute tool", async () => {
  const tool = "github";
  const action = "push";
  const params = { branch: "main" };
  const mutates = true;
  const approvalId = "appr-replay-001";
  const actionHash = hashAction(tool, action, mutates, params);

  const { url, close } = await startServer((req, res) => {
    if (req.method === "POST" && req.url === "/v1/authorize") {
      serveJSON(res, 200, {
        decision: "require_approval",
        action_hash: actionHash,
        approval: { approval_id: approvalId, action_hash: actionHash },
      });
    } else if (req.method === "GET" && req.url === `/v1/approvals/${approvalId}`) {
      serveJSON(res, 200, { status: "APPROVED", action_hash: actionHash });
    } else if (req.method === "POST" && req.url === `/v1/approvals/${approvalId}/consume`) {
      serveJSON(res, 409, { error: "already consumed" });
    } else {
      res.writeHead(404);
      res.end("unexpected");
    }
  });

  try {
    const client = newClient(url);
    let called = false;
    await assert.rejects(
      () =>
        protect(
          client,
          { tool, action, mutatesState: mutates, parameters: params },
          async () => {
            called = true;
            return "";
          },
          fastOpts
        )
    );
    assert.ok(!called, "tool must NOT be called when consume fails (409)");
  } finally {
    close();
  }
});

// ---------------------------------------------------------------------------
// 7. gateway unreachable on mutating action → error, tool NOT run
// ---------------------------------------------------------------------------

test("protect — unreachable gateway on mutating action: refuses and does NOT execute tool", async () => {
  const client = new AegisClient({
    baseUrl: "http://127.0.0.1:1", // nothing listening
    agentToken: "tok",
    tenantId: "tid",
    timeoutMs: 500,
  });
  let called = false;
  await assert.rejects(
    () =>
      protect(
        client,
        { tool: "fs", action: "delete_file", mutatesState: true, parameters: { path: "/etc/passwd" } },
        async () => {
          called = true;
          return "";
        },
        fastOpts
      )
  );
  assert.ok(!called, "tool must NOT be called when gateway is unreachable");
});

// ---------------------------------------------------------------------------
// 8. unknown decision → error (fail closed), tool NOT run
// ---------------------------------------------------------------------------

test("protect — unknown decision: fails closed, tool NOT run", async () => {
  const { url, close } = await startServer((_req, res) => {
    serveJSON(res, 200, { decision: "maybe" });
  });
  try {
    const client = newClient(url);
    let called = false;
    await assert.rejects(
      () =>
        protect(
          client,
          { tool: "github", action: "list_prs", mutatesState: false, parameters: {} },
          async () => {
            called = true;
            return "";
          },
          fastOpts
        )
    );
    assert.ok(!called, "tool must NOT be called on unknown decision");
  } finally {
    close();
  }
});

// ---------------------------------------------------------------------------
// 9. approval rejected → error, tool NOT run
// ---------------------------------------------------------------------------

test("protect — approval rejected: throws error, tool NOT run", async () => {
  const tool = "github";
  const action = "close_issue";
  const params = { issue: 99 };
  const mutates = true;
  const approvalId = "appr-rejected-001";
  const actionHash = hashAction(tool, action, mutates, params);

  const { url, close } = await startServer((req, res) => {
    if (req.method === "POST" && req.url === "/v1/authorize") {
      serveJSON(res, 200, {
        decision: "require_approval",
        action_hash: actionHash,
        approval: { approval_id: approvalId, action_hash: actionHash },
      });
    } else if (req.method === "GET" && req.url === `/v1/approvals/${approvalId}`) {
      serveJSON(res, 200, { status: "REJECTED", action_hash: actionHash, reason: "reviewer declined" });
    } else {
      res.writeHead(404);
      res.end("unexpected");
    }
  });

  try {
    const client = newClient(url);
    let called = false;
    await assert.rejects(
      () =>
        protect(
          client,
          { tool, action, mutatesState: mutates, parameters: params },
          async () => {
            called = true;
            return "";
          },
          fastOpts
        )
    );
    assert.ok(!called, "tool must NOT be called when approval is rejected");
  } finally {
    close();
  }
});

// ---------------------------------------------------------------------------
// 10. missing approval info in require_approval → fail closed
// ---------------------------------------------------------------------------

test("protect — require_approval without approval info: fails closed, tool NOT run", async () => {
  const { url, close } = await startServer((_req, res) => {
    serveJSON(res, 200, { decision: "require_approval" }); // no approval object
  });
  try {
    const client = newClient(url);
    let called = false;
    await assert.rejects(
      () =>
        protect(
          client,
          { tool: "github", action: "merge_pr", mutatesState: true, parameters: { pr: 7 } },
          async () => {
            called = true;
            return "";
          },
          fastOpts
        )
    );
    assert.ok(!called, "tool must NOT be called when approval info is missing");
  } finally {
    close();
  }
});

File: sdk-go/aegis/protect_test.go
package aegis_test

import (
	"context"
	"errors"
	"net/http"
	"net/http/httptest"
	"sync/atomic"
	"testing"

	"github.com/lavkushry/aegisagent/sdk-go/aegis"
	"github.com/lavkushry/aegisagent/sdk-go/canon"
)

// hashAction computes the expected action_hash for an AuthorizeRequest,
// mirroring what Protect does internally so tests can supply correct hashes.
// resource is a *string; we dereference to an untyped nil or string value so
// that canon (which does not handle pointer types) receives a supported type.
func hashAction(t *testing.T, req aegis.AuthorizeRequest) string {
	t.Helper()
	var resource any
	if req.Resource != nil {
		resource = *req.Resource
	}
	h, err := canon.CanonicalHash(map[string]any{
		"tool":          req.Tool,
		"action":        req.Action,
		"resource":      resource,
		"mutates_state": req.MutatesState,
		"parameters":    req.Parameters,
	})
	if err != nil {
		t.Fatalf("hashAction: %v", err)
	}
	return h
}

// toolCallTracker wraps a bool to record whether the tool was executed.
type toolCallTracker struct {
	called bool
}

func (tc *toolCallTracker) tool() error {
	tc.called = true
	return nil
}

// ---------- Protect — instant allow -----------------------------------------

func TestProtect_Allow_ExecutesTool(t *testing.T) {
	req := aegis.AuthorizeRequest{
		Tool:         "github",
		Action:       "list_prs",
		Parameters:   map[string]any{"repo": "aegis"},
		MutatesState: false,
	}

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		serveJSON(w, http.StatusOK, map[string]any{
			"decision":    "allow",
			"action_hash": hashAction(t, req),
		})
	}))
	defer srv.Close()

	client := newTestClient(srv.URL)
	tracker := &toolCallTracker{}

	err := aegis.Protect(context.Background(), client, req, func() error {
		return tracker.tool()
	})
	if err != nil {
		t.Fatalf("expected no error on allow, got: %v", err)
	}
	if !tracker.called {
		t.Error("expected tool to be called on allow decision")
	}
}

// ---------- Protect — deny --------------------------------------------------

func TestProtect_Deny_RefusesAndDoesNotExecute(t *testing.T) {
	req := aegis.AuthorizeRequest{
		Tool:         "github",
		Action:       "force_push",
		Parameters:   map[string]any{"branch": "main"},
		MutatesState: true,
	}

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		serveJSON(w, http.StatusOK, map[string]any{
			"decision": "deny",
			"reason":   "forbidden by policy",
		})
	}))
	defer srv.Close()

	client := newTestClient(srv.URL)
	tracker := &toolCallTracker{}

	err := aegis.Protect(context.Background(), client, req, func() error {
		return tracker.tool()
	})
	if err == nil {
		t.Fatal("expected error on deny, got nil")
	}
	if tracker.called {
		t.Error("tool must NOT be executed on deny")
	}
	var deniedErr *aegis.ErrDenied
	if !errors.As(err, &deniedErr) {
		t.Errorf("expected ErrDenied, got %T: %v", err, err)
	}
}

// ---------- Protect — require_approval → approved ---------------------------

func TestProtect_ApprovalPoll_PendingThenApproved_ExecutesTool(t *testing.T) {
	req := aegis.AuthorizeRequest{
		Tool:         "github",
		Action:       "merge_pr",
		Parameters:   map[string]any{"pr": float64(42)},
		MutatesState: true,
	}
	approvalID := "appr-poll-001"
	actionHash := hashAction(t, req)

	var pollCount int32

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch {
		case r.Method == http.MethodPost && r.URL.Path == "/v1/authorize":
			serveJSON(w, http.StatusOK, map[string]any{
				"decision":    "require_approval",
				"action_hash": actionHash,
				"approval": map[string]any{
					"approval_id": approvalID,
					"action_hash": actionHash,
				},
			})

		case r.Method == http.MethodGet && r.URL.Path == "/v1/approvals/"+approvalID:
			n := atomic.AddInt32(&pollCount, 1)
			if n < 2 {
				// First poll: still pending
				serveJSON(w, http.StatusOK, map[string]any{
					"status":      "PENDING",
					"action_hash": actionHash,
				})
			} else {
				// Second poll: approved
				serveJSON(w, http.StatusOK, map[string]any{
					"status":      "APPROVED",
					"action_hash": actionHash,
				})
			}

		case r.Method == http.MethodPost && r.URL.Path == "/v1/approvals/"+approvalID+"/consume":
			serveJSON(w, http.StatusOK, map[string]any{
				"action_hash": actionHash,
			})

		default:
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
			http.Error(w, "not found", http.StatusNotFound)
		}
	}))
	defer srv.Close()

	client := newTestClient(srv.URL)
	tracker := &toolCallTracker{}

	opts := aegis.ProtectOptions{
		PollInterval: 0, // zero → no sleep in tests
		MaxPolls:     5,
	}
	err := aegis.ProtectWithOptions(context.Background(), client, req, func() error {
		return tracker.tool()
	}, opts)
	if err != nil {
		t.Fatalf("expected approval to succeed, got: %v", err)
	}
	if !tracker.called {
		t.Error("tool must be called after approval")
	}
}

// ---------- Protect — hash mismatch → refuse --------------------------------

func TestProtect_HashMismatch_RefusesAndDoesNotExecute(t *testing.T) {
	req := aegis.AuthorizeRequest{
		Tool:         "github",
		Action:       "delete_branch",
		Parameters:   map[string]any{"branch": "feature-x"},
		MutatesState: true,
	}
	approvalID := "appr-mismatch-001"
	badHash := "000000000000000000000000000000000000000000000000000000000000dead"

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch {
		case r.Method == http.MethodPost && r.URL.Path == "/v1/authorize":
			// Return a mismatched action_hash in the approval
			serveJSON(w, http.StatusOK, map[string]any{
				"decision":    "require_approval",
				"action_hash": badHash,
				"approval": map[string]any{
					"approval_id": approvalID,
					"action_hash": badHash, // deliberately wrong
				},
			})

		default:
			// Should not reach polling or consume
			t.Errorf("unexpected request after hash mismatch: %s %s", r.Method, r.URL.Path)
			http.Error(w, "not found", http.StatusNotFound)
		}
	}))
	defer srv.Close()

	client := newTestClient(srv.URL)
	tracker := &toolCallTracker{}

	err := aegis.Protect(context.Background(), client, req, func() error {
		return tracker.tool()
	})
	if err == nil {
		t.Fatal("expected error on hash mismatch, got nil")
	}
	if tracker.called {
		t.Error("tool must NOT be called when action_hash mismatches")
	}
	var hashErr *aegis.ErrHashMismatch
	if !errors.As(err, &hashErr) {
		t.Errorf("expected ErrHashMismatch, got %T: %v", err, err)
	}
}

// ---------- Protect — hash mismatch on poll status → refuse -----------------

func TestProtect_HashMismatch_OnPollStatus_RefusesAndDoesNotExecute(t *testing.T) {
	req := aegis.AuthorizeRequest{
		Tool:         "github",
		Action:       "deploy",
		Parameters:   map[string]any{"env": "prod"},
		MutatesState: true,
	}
	approvalID := "appr-pollmismatch"
	correctHash := hashAction(t, req)
	tamperedHash := "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch {
		case r.Method == http.MethodPost && r.URL.Path == "/v1/authorize":
			serveJSON(w, http.StatusOK, map[string]any{
				"decision":    "require_approval",
				"action_hash": correctHash,
				"approval": map[string]any{
					"approval_id": approvalID,
					"action_hash": correctHash,
				},
			})

		case r.Method == http.MethodGet && r.URL.Path == "/v1/approvals/"+approvalID:
			// Gateway returns APPROVED but with a tampered hash (approve-then-swap attack)
			serveJSON(w, http.StatusOK, map[string]any{
				"status":      "APPROVED",
				"action_hash": tamperedHash,
			})

		default:
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
			http.Error(w, "not found", http.StatusNotFound)
		}
	}))
	defer srv.Close()

	client := newTestClient(srv.URL)
	tracker := &toolCallTracker{}

	opts := aegis.ProtectOptions{PollInterval: 0, MaxPolls: 3}
	err := aegis.ProtectWithOptions(context.Background(), client, req, func() error {
		return tracker.tool()
	}, opts)
	if err == nil {
		t.Fatal("expected error on tampered poll hash, got nil")
	}
	if tracker.called {
		t.Error("tool must NOT be called when poll action_hash mismatches")
	}
}

// ---------- Protect — unreachable gateway, mutating → refuse ----------------

func TestProtect_UnreachableGateway_MutatingAction_Refuses(t *testing.T) {
	client := aegis.NewClient(aegis.ClientOptions{
		BaseURL:    "http://127.0.0.1:1", // nothing listening
		AgentToken: "tok",
		TenantID:   "tid",
	})

	req := aegis.AuthorizeRequest{
		Tool:         "filesystem",
		Action:       "delete_file",
		Parameters:   map[string]any{"path": "/etc/passwd"},
		MutatesState: true,
	}
	tracker := &toolCallTracker{}

	err := aegis.Protect(context.Background(), client, req, func() error {
		return tracker.tool()
	})
	if err == nil {
		t.Fatal("expected error when gateway is unreachable on mutating action, got nil")
	}
	if tracker.called {
		t.Error("tool must NOT be called when gateway is unreachable for mutating action")
	}
}

// ---------- Protect — consume 409 → refuse (single-use replay defense) ------

func TestProtect_ConsumeFails_409_RefusesAndDoesNotExecute(t *testing.T) {
	req := aegis.AuthorizeRequest{
		Tool:         "github",
		Action:       "push",
		Parameters:   map[string]any{"branch": "main"},
		MutatesState: true,
	}
	approvalID := "appr-replay-001"
	actionHash := hashAction(t, req)

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch {
		case r.Method == http.MethodPost && r.URL.Path == "/v1/authorize":
			serveJSON(w, http.StatusOK, map[string]any{
				"decision":    "require_approval",
				"action_hash": actionHash,
				"approval": map[string]any{
					"approval_id": approvalID,
					"action_hash": actionHash,
				},
			})

		case r.Method == http.MethodGet && r.URL.Path == "/v1/approvals/"+approvalID:
			serveJSON(w, http.StatusOK, map[string]any{
				"status":      "APPROVED",
				"action_hash": actionHash,
			})

		case r.Method == http.MethodPost && r.URL.Path == "/v1/approvals/"+approvalID+"/consume":
			// Simulate already-consumed (replay attempt)
			serveJSON(w, http.StatusConflict, map[string]any{
				"error": "approval already consumed",
			})

		default:
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
			http.Error(w, "not found", http.StatusNotFound)
		}
	}))
	defer srv.Close()

	client := newTestClient(srv.URL)
	tracker := &toolCallTracker{}

	opts := aegis.ProtectOptions{PollInterval: 0, MaxPolls: 3}
	err := aegis.ProtectWithOptions(context.Background(), client, req, func() error {
		return tracker.tool()
	}, opts)
	if err == nil {
		t.Fatal("expected error when consume returns 409, got nil")
	}
	if tracker.called {
		t.Error("tool must NOT be called when consume fails")
	}
}

// ---------- Protect — approval rejected → refuse ----------------------------

func TestProtect_Rejected_Refuses(t *testing.T) {
	req := aegis.AuthorizeRequest{
		Tool:         "github",
		Action:       "close_issue",
		Parameters:   map[string]any{"issue": float64(99)},
		MutatesState: true,
	}
	approvalID := "appr-rejected-001"
	actionHash := hashAction(t, req)

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch {
		case r.Method == http.MethodPost && r.URL.Path == "/v1/authorize":
			serveJSON(w, http.StatusOK, map[string]any{
				"decision":    "require_approval",
				"action_hash": actionHash,
				"approval": map[string]any{
					"approval_id": approvalID,
					"action_hash": actionHash,
				},
			})

		case r.Method == http.MethodGet && r.URL.Path == "/v1/approvals/"+approvalID:
			serveJSON(w, http.StatusOK, map[string]any{
				"status":      "REJECTED",
				"action_hash": actionHash,
				"reason":      "reviewer declined",
			})

		default:
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
			http.Error(w, "not found", http.StatusNotFound)
		}
	}))
	defer srv.Close()

	client := newTestClient(srv.URL)
	tracker := &toolCallTracker{}

	opts := aegis.ProtectOptions{PollInterval: 0, MaxPolls: 3}
	err := aegis.ProtectWithOptions(context.Background(), client, req, func() error {
		return tracker.tool()
	}, opts)
	if err == nil {
		t.Fatal("expected error on rejection, got nil")
	}
	if tracker.called {
		t.Error("tool must NOT be called when approval is rejected")
	}
}

// ---------- Protect — missing approval info → refuse (fail closed) -----------

func TestProtect_RequireApproval_MissingApprovalInfo_FailsClosed(t *testing.T) {
	req := aegis.AuthorizeRequest{
		Tool:         "github",
		Action:       "merge_pr",
		Parameters:   map[string]any{"pr": float64(7)},
		MutatesState: true,
	}

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Gateway says require_approval but does not include the approval object
		serveJSON(w, http.StatusOK, map[string]any{
			"decision": "require_approval",
			// no "approval" key
		})
	}))
	defer srv.Close()

	client := newTestClient(srv.URL)
	tracker := &toolCallTracker{}

	err := aegis.Protect(context.Background(), client, req, func() error {
		return tracker.tool()
	})
	if err == nil {
		t.Fatal("expected error when approval info missing, got nil")
	}
	if tracker.called {
		t.Error("tool must NOT be called when approval info is missing")
	}
}

// ---------- Protect — unknown decision → refuse (fail closed) ---------------

func TestProtect_UnknownDecision_FailsClosed(t *testing.T) {
	req := aegis.AuthorizeRequest{
		Tool:         "github",
		Action:       "list_prs",
		Parameters:   map[string]any{},
		MutatesState: false,
	}

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		serveJSON(w, http.StatusOK, map[string]any{
			"decision": "maybe", // not a recognised decision
		})
	}))
	defer srv.Close()

	client := newTestClient(srv.URL)
	tracker := &toolCallTracker{}

	err := aegis.Protect(context.Background(), client, req, func() error {
		return tracker.tool()
	})
	if err == nil {
		t.Fatal("expected error on unknown decision, got nil")
	}
	if tracker.called {
		t.Error("tool must NOT be called on unknown decision")
	}
}
